[weiseng18] iP

weiseng18 opened the pull request:
## Panorama

> "Like a bird 🐦 soaring high, a panorama gives you a bird's eye view 🌁 to see the big picture and plan your next move." - weiseng18 

**Panorama** is a todo list app that helps you to manage your tasks efficiently. It

- allows you to take a step forward (by choosing your next task), even if you don't know where you should go
- allows you to imagine different views (with the `find` command) and appreciate the scenery

```java
public class PanoramaJourney {
    public static void main(String[] args) {
        int steps = 0;
        
        while (true) {
            steps++;
            if (steps % 100 == 0) {
                System.out.println("You've walked " + steps + " steps, take a breather 🌅");
            }
            // Remember, it's okay to take a break and reflect
            if (steps == 1000) {
                System.out.println("You reached a new milestone! Look at the view 👀");
                break;
            }
        }
    }
}
```

## inspiration

1. [端程山 (Panorama) by MyGO!!!!!](https://www.youtube.com/watch?v=1c2uSrAGF9Q&ab_channel=MyGO%21%21%21%21%21)
2. my Anki flashcard for 鳥瞰図 (bird's eye view)

## obligatory checklist

- [x] a heading
- [x] a bullet list
- [x] a numbered list
- [x] a fenced code block (with syntax highlighting)
- [x] a task list
- [x] an emoji
- [x] a blockquote
- [x] a hyperlink
- [x] inline code
- [x] some text formatting: bold, italic, strikethrough, etc.

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

weiseng18 committed:
rename chatbot to Panorama

weiseng18 committed:
initial skeletal version

weiseng18 committed:
implement echo and exit

weiseng18 committed:
refactor: move constant strings out

weiseng18 committed:
refactor: put code into functions

weiseng18 committed:
create list

weiseng18 committed:
implement level 2

weiseng18 committed:
add *.class to gitignore

weiseng18 committed:
create build script

weiseng18 committed:
add Task class with done status

weiseng18 committed:
display done status

weiseng18 committed:
can mark/unmark tasks

weiseng18 committed:
add output to indicate marking/unmarking of tasks

weiseng18 committed:
refactor to create abstract class Task

weiseng18 committed:
handle deadline command

weiseng18 committed:
handle event command

weiseng18 committed:
stadndardize indent for output of adding task

weiseng18 committed:
basic automated testing

weiseng18 committed:
handle empty todo description and unknown command exceptions

weiseng18 committed:
basic delete command

weiseng18 committed:
throw empty description exception for deadline and todo

weiseng18 committed:
display command help

weiseng18 committed:
add final keyword to constants

weiseng18 committed:
implement basic load/save

weiseng18 committed:
add sample data.txt

weiseng18 committed:
Merge branch 'branch-Level-7'

weiseng18 committed:
Use defined constant instead of hardcoded string

weiseng18 committed:
Refactor into packages

weiseng18 committed:
Capitalize separator and indent since they are constants

weiseng18 committed:
Refactor task handling into TaskManager class

weiseng18 committed:
Use Command enum

weiseng18 committed:
Fix Java String comparison issue

weiseng18 committed:
Merge branch 'refactor-after-level-7'

weiseng18 committed:
Implement basic date handling

weiseng18 committed:
Merge branch 'branch-Level-8'

weiseng18 committed:
Rename TaskManager -> TaskList

weiseng18 committed:
Extract Storage from TaskList

weiseng18 committed:
Extract Ui functionality into own class

weiseng18 committed:
Move Task input parsing to Parser

weiseng18 committed:
Merge branch 'A-MoreOOP'

weiseng18 committed:
Made duke package

weiseng18 committed:
Merge branch 'A-Packages'

weiseng18 committed:
Merge branch 'add-gradle-support'

weiseng18 committed:
Update build.gradle to match my setup

weiseng18 committed:
Move data.txt to root

weiseng18 committed:
Remove unneeded run.sh

weiseng18 committed:
Merge branch 'A-Gradle'

yixianggg commented on src/main/java/duke/MyTask/Task.java:
```diff
@@ -0,0 +1,41 @@
+public abstract class Task {
+    String name;
+    boolean isDone;
+    char c;
```
Perhaps instead of naming the char 'c', could name it something else like taskType?

yixianggg commented on src/main/java/duke/Panorama.java:
```diff
@@ -0,0 +1,109 @@
+
+// Exceptions folder
+import duke.MyException.EmptyDescriptionException;
+import duke.MyException.UnknownCommandException;
```
I like how you split up the imports neatly! 👍

yixianggg commented on src/main/java/duke/Panorama.java:
```diff
@@ -0,0 +1,109 @@
+                break;
+            default:
+                throw new UnknownCommandException();
+        }
```
I believe based on the coding standards there shouldn't be indentations for case 😄

weiseng18 replied:
thanks!! rectified

will probably fix the rest as I work on `A-CodingStandard`

yixianggg commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,73 @@
+        Scanner myReader = new Scanner(myFile);
+
+        // Temporary list for return
+        List<Task> memory = new ArrayList<>();
```
Perhaps could name the list something more descriptive of what it stores?

yixianggg commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,71 @@
+    public static final String SEPARATOR = "    ____________________________________________________________";
+    public static final String INDENT = "     "; // 5 spaces
+
+    static void welcome_greeting() {
```
Perhaps could name method to be an action and use camelCase i.e. printWelcomeGreeting

Similar for exit_greeting method!

yixianggg reviewed (commented):
LGTM. Just some nits to take note of! 😄

weiseng18 committed:
Setup JUnit testing

weiseng18 committed:
Add test for Task subclasses

weiseng18 committed:
Add test for DateParser.parse

weiseng18 committed:
Add test for Parser.parseTodoInput

weiseng18 committed:
Merge branch 'A-JUnit'

Ma-Xizi commented on src/main/java/duke/MyTask/Task.java:
```diff
@@ -0,0 +1,41 @@
+        this.c = c;
+    }
+
+    public void setIsDone(boolean status) {
```
According to Java coding standards, maybe setDone is a better name?

Ma-Xizi commented on src/main/java/duke/Panorama.java:
```diff
@@ -0,0 +1,109 @@
+
+        // Save task list to file after exiting
+        try {
+            List<Task> memory = taskList.getMemory();
```
Maybe a better name could be chosen for this list of tasks, eg, tasksInMemory? This also applies to attribute name in TaskList class.

Ma-Xizi commented on src/main/java/duke/MyException/EmptyDescriptionException.java:
```diff
@@ -0,0 +1,8 @@
+
+public class EmptyDescriptionException extends Exception {
+    public EmptyDescriptionException() {
+        super("     Description cannot be empty.");
```
Is there any better way to add the indent?

weiseng18 replied:
yup, I intend to expose a static constant from the main class which I can refer to!

Ma-Xizi commented on src/main/java/duke/Parser/DateParser.java:
```diff
@@ -0,0 +1,34 @@
+     * Returns a date string in the format MMM dd yyyy HHmm.
+     * For output to terminal.
+     */
+    public static String format(LocalDateTime dateTime) {
+        return dateTime.format(OUTPUT_FORMATTER);
+    }
+
+    /**
+     * Returns a date string in the format yyyy-mm-dd HHmm.
+     * For output to hard disk.
+     */
+    public static String formatForHardDisk(LocalDateTime dateTime) {
+        return dateTime.format(INPUT_FORMATTER);
+    }
+}
```
Maybe can give 'format' a better name to distinguish 'format' and 'formatForHardDisk' more

weiseng18 replied:
good pt! thanks!

Ma-Xizi reviewed (commented):
Looks good.
Some cosmetic changes recommended.

weiseng18 committed:
Remove trailing whitespaces

weiseng18 committed:
Fix indentation for switch block

weiseng18 committed:
Mark variables as explicitly private

weiseng18 committed:
Remove initialization tests as they only check the validity of private attributes

weiseng18 committed:
Make Command constructor private

weiseng18 committed:
Add Javadoc for Command.java

weiseng18 committed:
Add Javadoc for Ui.java

weiseng18 committed:
Add Javadoc for abstract class Task and its subclasses

weiseng18 committed:
Add Javadoc for Parser package

weiseng18 committed:
Fix remaining Javadoc issues

weiseng18 committed:
Merge branch 'A-CodingStandard'

weiseng18 committed:
Implement Level 9

weiseng18 committed:
Merge branch 'Level-9'

weiseng18 committed:
Update handleCommand Javadoc to include FIND

weiseng18 committed:
Use Ui in a static way

weiseng18 committed:
Rename c -> taskType

weiseng18 committed:
Update Ui method names to follow convention

weiseng18 committed:
Use Ui.INDENT constant

weiseng18 committed:
Rename setIsDone -> setDone to follow conventions

weiseng18 committed:
Rename test functions to make the exception being tested clear

weiseng18 committed:
Use Ui.INDENT

weiseng18 committed:
Add test for Command class

weiseng18 committed:
Make return types more specific for Parser.java

weiseng18 committed:
Move Task input parsing out of TaskList class

weiseng18 committed:
Move Task ID parsing out of TaskList class

weiseng18 committed:
Update addTask methods

weiseng18 committed:
Remove printTaskAdded since it is only called in one location

weiseng18 committed:
Refactor handleCommand to be in its own class

weiseng18 committed:
Create Message class to contain hardcoded strings

weiseng18 committed:
Use constants from Message class

weiseng18 committed:
Remove indent from the Exception itself and make the main class print it instead

weiseng18 committed:
Remove unused functions

weiseng18 committed:
Move command formats into Message

weiseng18 committed:
Create individual functions for commands with input handling

weiseng18 committed:
Correctly use first parameter as expected value in tests

weiseng18 committed:
Add happy flow unit test for mark and unmark

weiseng18 committed:
Add happy flow unit test for delete

weiseng18 committed:
Add happy flow unit test for task creation functions

weiseng18 committed:
Add checkstyle

weiseng18 committed:
Add final keyword to constants in Message.java

weiseng18 committed:
Fix package import order checkstyle issues

weiseng18 committed:
Remove unused variable

weiseng18 committed:
Break long lines

weiseng18 committed:
Add Javadoc for Message class

weiseng18 committed:
Update Javadoc to the latest variable name

weiseng18 committed:
Fix variable name convention

weiseng18 committed:
Move constant strings to Message

weiseng18 committed:
Split variable declarations onto different lines

weiseng18 committed:
Make CommandHandlerOutput attributes private

weiseng18 committed:
Fill up missing Javadoc

weiseng18 committed:
Add JavaFX to Gradle

weiseng18 committed:
Duplicate Task class with minor changes

weiseng18 committed:
Reimplement TaskList

weiseng18 committed:
Reimplement Command interface with task creation commands

weiseng18 committed:
Implement DeleteCommand

weiseng18 committed:
Implement MarkCommand and UnmarkCommand

weiseng18 committed:
Reimplement Storage

weiseng18 committed:
Implement Ui

weiseng18 committed:
Duplicate Message class

weiseng18 committed:
Duplicate custom exception classes

weiseng18 committed:
Add COMMAND_WORD attribute to Command related classes

weiseng18 committed:
Implement Parser that handles add commands

weiseng18 committed:
Add Delete, Mark, Unmark

weiseng18 committed:
Add List

weiseng18 committed:
Add isBye function to Command interface

weiseng18 committed:
Introduce ByeCommand

weiseng18 committed:
Add Panorama class

weiseng18 committed:
Update gradle entrypoint

weiseng18 committed:
Remove time in data.txt

weiseng18 committed:
Add missing ByeCommand to Parser

weiseng18 committed:
Remove trailing endline for ListCommand

weiseng18 committed:
Implement FindCommand

weiseng18 committed:
Fix Javadoc for Storage

weiseng18 committed:
Implement saving

weiseng18 committed:
Add endline in output for readability

weiseng18 committed:
Add test for DateParser

weiseng18 committed:
Add test for Task classes

weiseng18 committed:
Rename test functions to follow naming convention

weiseng18 committed:
Add test for Parser

weiseng18 committed:
Remove irrelevant tests

weiseng18 committed:
Fix parsing bug in EventCommand

weiseng18 committed:
Add parse test for Deadline and Event

weiseng18 committed:
Remove last irrelevant test

weiseng18 committed:
Remove duke

weiseng18 committed:
Add happy flow test for mark, unmark, delete

weiseng18 committed:
Add Help command

weiseng18 committed:
Remove unused import

weiseng18 committed:
Move static constants to top of class definition

weiseng18 committed:
Fix Javadoc issues

weiseng18 committed:
Setup GUI

weiseng18 committed:
Add cosmetic changes

weiseng18 committed:
Display welcome message in GUI

weiseng18 committed:
Update Storage class to interface with List<Task> rather than TaskList

weiseng18 committed:
Move Storage into TaskList for inline saving

weiseng18 committed:
Remove unused imports

weiseng18 committed:
Fix checkstyle

weiseng18 committed:
A-Assertions (#1)

* Add assert to ensure that storage is not null in TaskList

This assertion is intended to catch situations where the storage field
may be null. The assertion is added just before attempting to use the
storage field like in storage.saveTasks() or storage.loadTasks().

Given that Storage is a custom class, this additional check is useful to
catch potential bugs in my implementation.

* Add assert to ensure that taskList is not null in relevant Commands

Certain Commands contain a reference to taskList, and mutate the list.
Having the assert here helps to spot bugs if somehow the taskList is
null.

weiseng18 committed:
A-CodeQuality (#2)

* Remove unused function

* Add else branch to handle invalid input for loadTasks()

weiseng18 committed:
A-CI (#3)

* Add sample gradle.yml file

* Fix import order in tests

* Fix visibility modifier issues in tests

weiseng18 committed:
C-FriendlierSyntax (#4)

* Add COMMAND_SHORTHAND constant for all Commands

* Update Parser to allow shorthands as defined in COMMAND_SHORTHAND per Command

* Update help page to indicate that preset shorthands are allowed

weiseng18 committed:
Rename boolean found -> isFound to match naming convention

weiseng18 committed:
Merge pull request #6 from weiseng18/branch-A-Assertions

Rename boolean found -> isFound to match naming convention

weiseng18 committed:
Add Jacoco

weiseng18 committed:
Update workflow to include Codecov

weiseng18 committed:
Merge pull request #7 from weiseng18/branch-A-CodeQuality

Include Codecov

weiseng18 committed:
A-MoreTesting (#8)

* Refactor test to use COMMAND_WORD instead of hardcoded constant

* Add tests for using shorthand alias

* Add isTestEnv attribute to disable read/write from file in a test env

* Add missing instantiation

* Fix constructor parameter

* Add basic tests for TaskList

* Fix checkstyle

* Make isTestEnv required param

* Pass isTestEnv to find() result

* Add test for mark, unmark, delete

* Update delete test to check that returned Task is correct

* Add test for find

* Add test for the remaining Commands in Parser

weiseng18 committed:
Handle more exceptions (#9)

* Handle missing or noninteger ID in user input

* Handle ID being out of bounds of the task list (1-indexed)

* Handle empty search keyword for find command

* Return default string in toString() when task list is empty

* Modify execute return type from String -> Response to implement bye properly

* Implement custom exception for date parsing

* Implement PanoramaException base class

* Handle invalid usage for Deadline and Event Commands

weiseng18 committed:
Add Ui.png

weiseng18 committed:
Update user guide

weiseng18 committed:
Add missing backticks in README

weiseng18 committed:
Update README.md

weiseng18 committed:
Rename default tasks

weiseng18 committed:
Return empty list if no previous data found
